### PR TITLE
33 Previews and extra data for resources

### PR DESCRIPTION
**Pull request by kirillt**
`ResourceMetaExtra` substructure of `ResourceMeta` introduced. This field is used to store dynamically-determined data which depends on `kind` of `ResourceMeta` (like video, documents, images etc.)

Previews are implemented using global singleton object and their generation is triggered during indexing. This has to be changed in future (see #115). We store generated previews in cache folder so far (this will be changed upon implementation of persisted/replicated cache). For images we don't generate previews, but just use themselves. For any kind of resource, besides preview, thumbnail is also generated which is preview scaled down to 72x108 and stored in separate folder. Existence of preview/thumbnail for given resource is done prepending constant path to its `id`.

What needs to be finished prior to merge into main:
* Room structures for storing `ResourceMetaExtra` must be verified: `ResourceExtra` and `ResourceWithExtra`
* Previews and thumbnails generation should be verified as well
* Extra text snippets should be overlayed for videos and PDFs (see method `draw()` in `VideoMetaExtra` and `DocumentMetaExtra`)

Also would be good:
~~* Previews for videos (#33) is what gave name to this PR, but we could finish it later actually~~
~~* Make sortings in resources grid use `ResourceMeta` fields `size` and `modified` instead of querying file system~~
~~* Verify coroutines for loading thumbnails (see usage of `ThumbnailTarget`)~~

### CODE REVIEW

**Comment by mdrlzy** [comment]
Glide sometimes loading wrong image on RecyclerView. 
Related issue: https://github.com/bumptech/glide/issues/2847
<p align="center">
 |

</p>

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/extra/DocumentMetaExtra.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        setOf("pdf", "txt", "doc", "docx", "odt", "ods", "md")
+
+    fun extract(path: Path): ResourceMetaExtra? = null
+
+    fun draw(extra: ResourceMetaExtra, pagesTV: TextView, verbose: Boolean) {
+        val pages = extra.data[MetaExtraTag.PAGES]
+        if (pages != null) {
+            val label = if (verbose) "$pages pages" else "$pages"
+            pagesTV.textOrGone(label)
+        }
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/extra/DocumentMetaExtra.kt:20]: Maybe the text generation can be left here, but the work with the view should be in the `ui` package
Same for `ImageMetaExtra` and `VideoMetaExtra`

File: app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewGenerators.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import android.graphics.Bitmap
+import space.taran.arknavigator.ui.fragments.preview.generator.PdfPreviewGenerator
+import java.nio.file.Path
+
+object PreviewGenerators {
+    var BY_EXT: Map<String, (Path) -> Bitmap> = mapOf(
+        "pdf" to { path: Path -> PdfPreviewGenerator.generate(path) }
+        //todo moar
+    )
+}
```
Comment (mdrlzy) [inline app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewGenerators.kt:12]: I think preview generation should be in `model` package

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/extra/DocumentMetaExtra.kt (hunk 1 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/extra/DocumentMetaExtra.kt:20]: Yes, you are right. Feel free to change this.

File: app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewGenerators.kt (hunk 2 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewGenerators.kt:12]: Agree :+1:

File: app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewAndThumbnail.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                preview = previewPath(resource.id),
+                thumbnail = thumbnail)
+        }
+
+        fun forget(id: ResourceId) {
+            Files.deleteIfExists(previewPath(id))
+            Files.deleteIfExists(thumbnailPath(id))
+        }
+
+        @ExperimentalCoroutinesApi
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewAndThumbnail.kt:72]: I guess we don't need it anymore

File: app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewAndThumbnail.kt [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        private const val THUMBNAIL_HEIGHT = 128
+        private const val COMPRESSION_QUALITY = 100
+
+        private fun previewPath(id: ResourceId): Path =
+            PREVIEWS_STORAGE.resolve(id.toString())
+        private fun thumbnailPath(id: ResourceId): Path =
+            THUMBNAILS_STORAGE.resolve(id.toString())
+
+        init {
+            if (PREVIEWS_STORAGE.notExists()) Files.createDirectories(PREVIEWS_STORAGE)
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/ui/fragments/preview/PreviewAndThumbnail.kt:45]: Actually `Files.createDirectories` creates the directory if it doesn't exist.
That distinguishes this method from `Files.createDirectory` if I remember correctly.

**Pull request merged**